proposals/permissions-requestSiteAccess-api.md

```diff
@@ -0,0 +1,135 @@
+# permissions.requestSiteAccess() API
+
+## API Overview
+
+### Background
+
+Extensions can request host permissions via:
+- Manifest declared `host_permissions`, granted at install time by default
+- Manifest declared `optional_host_permissions`, granted at runtime after a user gesture by `<browser>.permissions.request()`
+- Manifest declared `matches` for a `content_script`, granted at install time by default
+
+Users can withhold host permissions, causing the extension to only have access when the extension is [invoked](https://developer.chrome.com/docs/extensions/develop/concepts/activeTab#invoking-activeTabhttps://developer.chrome.com/docs/extensions/develop/concepts/activeTab#invoking-activeTab).
+
+Each browser may decide how it signals the user when an extension is requesting access (e.g through the extensions menu). However, there is no way for an extension to explicitly signal at runtime it’s requesting site access after it was withheld without a user gesture and a heavy-weight permission dialog.
+
+
+### Objective
+
+Allow the extension to show site access requests at runtime without any user gesture in a less-obtrusive way than with permissions.request(). This can be done with a new API that:
+
+- Applies to a specific tab or document id
+- Doesn’t need to be made inside the handler for a user action
+- Shows the request in the UI, handled differently by each browser. See more in [UI Elements and User-Visible Effects](#ui-elements-and-user-visible\-effects) section
+- When accepted, grants always access to the site’s origin
+- Resets the request on cross-origin navigation
+
+#### Use Cases
+
+An extension requested access to a site but the user withheld its access and forgot about it. Extension wants to signal it needs site access to execute. For example, “shopping” extension wants to show site access when user navigates to “amazon.com” and access was withheld.
+
+### Consumers
+
+Extensions that want to signal the user when they need site access.
+
+## Implementation
+
+### API Schema
+
+```
+// Requests access to a specific tab or document. Access request will only be
+// signaled to the user if requesting access to a site the extension can be
+// granted access to (i.e., one specified in host_permissions,
+// optional_host_permissions, a content script match pattern, or an applicable
+// activeTab site).
+// Only one of tabId or documentId must be declared. Request will be
+// automatically reset on navigation.
+<browser>.permissions.requestSiteAccess(
+  // The id of the tab where site access requests can be shown.
+  number?: tabId,
+  // The id of a document where site access requests can be shown.
+  string?: documentId,
+  callback?: function
```

Comment:
Should the `Promise` / `callback` be passed a `bool`? What happens if the request is denied?

Reply:
> Does this also return a Promise? 
Yes, this should return a promise. In Chrome, promises are supported in Manifest V3 and later, but callbacks are provided for backward compatibility. You cannot use both on the same function call. The promise resolves with the same type that is passed to the callback.

> Should the Promise / callback be passed a bool? 
As for the return, there are a couple options:
  a. Returns a boolean whether the request was valid (e.g request will have no effect if extension doesn't have host permissions)
  b. Returns a boolean whether the user accepted the request

I think b. makes sense. One detail is that on cross-origin navigation, the request is effectively not accepted so it would need to return false. Haven't looked at the code, but there should be a way to do so.

Will update the doc once we decide on return value

> What happens if the request is denied?
We should probably add some limitation on cannot re-request on a denied request. Otherwise, it would be very intrusive to the user

Reply:
Taking a second look, I think it's better to return a. whether the request is valid. My reasoning is better explained in the proposal (see other alternatives considered). Basically, to have a more clear difference vs showing a permission request and granting a permission (which can be done in various ways).

However, happy to discuss this one further

Reply:
As discussed on _minutes/2024-02-29-wecg.md, returning whether request is valid is preferred. Resolving this comment